### DISCUSSION TITLE
Refactor storing answer in localstorage

### DISCUSSION THREAD

**Discussion by shrihari1999** [octokatherine/word-master#126] [Ideas]
Since wordle comes up with a new word every day, you cannot write an infinite loop to keep solving the puzzle. But since Word Master allows you to play again, one can find the answer in localStorage, and write an infinite loop to keep solving the puzzle.

Demonstration: https://user-images.githubusercontent.com/13913674/154923933-0c0a6431-8bd8-43ba-95d8-5d0ad18ca6f9.mp4

Anything we can do to prevent this?? 🤔🤔

**Comment by octokatherine** [comment]
I'm really not worried about this to be honest. It's very easy to cheat at this game if you want to, you wouldn't even do an infinite loop of solving the game because you could just directly update the streak if you wanted to

**Comment by LWChris** [comment]
As long as there is no global leaderboard, there's no point in trying to prevent cheating. If you want to cheat, you can always do it. You can always go into debug mode and either edit the local storage, or even just the source code for a screen shot. It's like buying a lockable fridge to keep yourself from eating your own ice cream. If you want to destroy your own honest achievements by cheating them, go ahead and do it. 🤷‍♂️

**Comment by shrihari1999** [comment]
I'll try to explain my POV.

1. It's not about the streak. It's about wanting to the play the game, knowing that the answer is right there in your local storage for me to see.

2. https://wordleunlimited.org/ has a similar website. They don't expose the answer directly. Again, it's not about the streak. It's about allowing users to take your game seriously. If you know that the answer is RIGHT THERE, you're not gonna spend time thinking.

 Knowing you can cheat makes a lot of difference when there are sites where you can't cheat.

I didn't respond since Katherine said 'shes not worried about cheating'. Subjective, i ignored

But you're saying 'theres no point in preventing cheating' -- implies an objective fact... Had to disagree

**Reply by LWChris** [reply]
Okay, one question though - would it be enough for you if the answer was slightly encoded, say if it was stored in some numerical way that if you look at it isn't immediately obvious what the word is?

**Reply by shrihari1999** [reply]
Sure, that's what https://wordplay.com/daily does using a 'Game ID', but they show it to the user. Note that this ID cannot be seen in local storage.

Not showing the Game ID would do the job.

**Reply by LWChris** [reply]
Just because it is not in the local storage does not mean it's not accessible for a script. If you are dedicated to cheat, you can cheat. That's the point I was making. I can totally work out what the right solution is for WordPlay using scripts. Would take me some time since the code is minified which is always a pain to analyze, but I know it's stored in some variable for some JavaScript object.

Also, WordPlay does not seem to offer picking up where you left off, right? That is kind-of the crux here. If you want the user to be able to close the browser or refresh the page and get the same puzzle with the same progress, you have to save some data that also encodes what the final answer should be.

So, if you want to protect the solution from being obvious in the local storage, that's a can-do. If you want the solution to be not stored on the user-side, that's not possible, because there's no backend side to this game to store that answer server-side for you.

In other words - I can easily devise a mechanism to encode the answer. Then, in the local storage, you'd see "answer = {some big number}". It's something I had considered developing and suggesting, together with a feature to create a share link to any puzzle. If that's enough for you, it can be done. If you are going to argue that one can extract the code on how to convert that number back into the right answer from the source code, then you're out of luck, because this will always be possible necessarily, because

* Some information has to be on the client's computer that can identify the answer
* Some code has to be on the page to restore the answer from that information
* Both will be accessible from the client side, since this is a JavaScript app without server-side storage

**Comment by LWChris** [comment]
This thread was just bumped by a spam bot (I think).

But I'll take this opportunity to announce that I have actually some exciting news to share on this one: I have devised a way to encode every possible 5-letter-combination into a number that is guaranteed to be within the range of JavaScript's native "number" type, reversible into these 5 letters as well, and constructed such that the human mind cannot see or estimate from experience with other word-number pairs what the solution might be.

## Examples
### Close values &#8655; similar words
BLOOM &lrarr; 2.895.986.445.493
HEAVY &lrarr; 2.896.404.173.609
ADOPT &lrarr; 2.897.686.793.147

### Similar words &#8655; close values
PILED &lrarr; 2.319.866.553.677
PILES &lrarr; 3.096.030.084.647
PILLS &lrarr; 3.575.857.625.441

### Larger values &#8655; higher alphabetical sort order
BACON &lrarr; 1.892.432.892.527
AUDIO &lrarr; 2.486.751.304.967
LIMIT &lrarr; 3.126.114.671.843
HUNTS &lrarr; 4.603.453.640.837
PROPS &lrarr; 5.142.307.042.591
STUMP &lrarr; 5.405.776.446.131
LOUSY &lrarr; 5.607.483.773.999

### Higher alphabetical sort order &#8655; larger values
AGONY &lrarr; 3.123.406.018.429
CLUNK &lrarr; 3.160.802.066.767
FAVOR &lrarr; 3.134.709.548.413
PIGGY &lrarr; 3.187.133.400.907
WIDTH &lrarr; 3.135.242.763.371

It's convoluted code, and the result is not perfect, but it does a GREAT job for exactly this purpose I think. It could also be used to implement a "Share" feature, where you can generate a link like `/word-master/?id=2628683176591` to share with your friends to have them try exactly the same puzzle as you are currently solving without giving away the solution.

And the cherry on the icing: since it is all just a (very convoluted, but still) basic mathematical algorithm and not for example a hard-coded dictionary key per answer, adding or removing words from the pool of solutions doesn't break anything. No need to come up with new unique ids every time you add a new word to the list of answers. But also no worries that loading ancient game states might break or months old links will stop working after that particular solution was removed from that list of answers in live.

@octokatherine I'd love to keep working on this and other features, and eventually merge them back into your repo. But maintaining independently acceptable pull-requests for such big features that all build on each other (like the hints and my extreme mode) gets increasingly difficult. Please tell me if there are any plans on your side to keep this game updates, otherwise I'll "secede" my fork from yours and implement these advanced features to my liking in my own copy from now on and stop bothering you with merge requests. 🙂 

(Edit: I just realized I copied all those numeric values from the output of an old implementation (my first draft); but the points are still true for the new version that I have actually implemented in my local code base.)

**Reply by octokatherine** [reply]
@LWChris I think I would opt to let you go ahead and just maintain that in your own fork and keep the complexity here down because I don't have a lot of time to maintain it currently. Thanks though and good luck, it sounds cool!
